### PR TITLE
fix: move ctor for command help class to its own function

### PR DESCRIPTION
**Pull request by peternhale**
@W-9708749@

### CODE REVIEW

**Review by mdonnalley — APPROVED** [review]

File: src/main.ts [hunk 1]
```diff
@@ -54,6 +55,7 @@ export async function run(argv = process.argv.slice(2), options?: Interfaces.Loa
     argv = argv.filter(arg => !getHelpFlagAdditions(config).includes(arg))
     const Help = await loadHelpClass(config)
     const help = new Help(config, config.pjson.helpOptions)
```
Comment (peternhale) [inline src/main.ts:57]: I did not want to add another parameter to the constructor, so capturing original argv and then setting it directly o n the Help class instance.

File: src/main.ts [hunk 2]
```diff
@@ -30,6 +30,7 @@ export const versionAddition = (argv: string[], config?: Interfaces.Config): boo
 }
 
 export async function run(argv = process.argv.slice(2), options?: Interfaces.LoadOptions) {
+  const originalArgv = [...argv]
```
Comment (peternhale) [inline src/main.ts:33]: Capture argv as passed to run

File: src/main.ts [hunk 3]
```diff
@@ -54,6 +55,7 @@ export async function run(argv = process.argv.slice(2), options?: Interfaces.Loa
     argv = argv.filter(arg => !getHelpFlagAdditions(config).includes(arg))
     const Help = await loadHelpClass(config)
     const help = new Help(config, config.pjson.helpOptions)
+    help.originalArgv = originalArgv
```
Comment (peternhale) [inline src/main.ts:58]: set the value on the help class.

File: src/help/index.ts [hunk 4]
```diff
@@ -30,11 +30,21 @@ function getHelpSubject(args: string[], config: Interfaces.Config): string | und
...
 
 export abstract class HelpBase extends HelpFormatter {
+  private argvCopy: string[] = []
+
   constructor(config: Interfaces.Config, opts: Partial<Interfaces.HelpOptions> = {}) {
     super(config, opts)
     if (!config.topicSeparator) config.topicSeparator = ':' // back-support @oclif/config
   }
 
+  public get originalArgv(): string[] {
```
Comment (peternhale) [inline src/help/index.ts:40]: getter/setter

**Comment by RodEsp** [comment]
Review comments captured in slack: https://platformcloud.slack.com/archives/C01U75RNU1G/p1631567124234900?thread_ts=1631553503.228700&cid=C01U75RNU1G

File: src/help/index.ts [hunk 5]
```diff
@@ -90,6 +90,8 @@ export class Help extends HelpBase {
   }
 
   public async showHelp(argv: string[]) {
+    argv = argv.filter(arg => !getHelpFlagAdditions(this.config).includes(arg))
```
Comment (peternhale) [inline src/help/index.ts:93]: Moved the stripping of help flags to showHelp from main.ts.

Other function call via showHelp expect the help flags to be removed.

File: src/main.ts [hunk 6]
```diff
@@ -51,7 +51,6 @@ export async function run(argv = process.argv.slice(2), options?: Interfaces.Loa
 
   // display help version if applicable
   if (helpAddition(argv, config)) {
-    argv = argv.filter(arg => !getHelpFlagAdditions(config).includes(arg))
```
Comment (peternhale) [inline src/main.ts:54]: main.ts should not have any knowledge of what showHelp expects in argv, just pass the whole thing as is.

File: src/help/index.ts [hunk 7]
```diff
@@ -202,10 +204,14 @@ export class Help extends HelpBase {
       command.id = command.id.replace(/:/g, this.config.topicSeparator)
       command.aliases = command.aliases && command.aliases.map(a => a.replace(/:/g, this.config.topicSeparator))
     }
-    const help = new this.CommandHelpClass(command, this.config, this.opts)
+    const help = this.getCommandHelpClass(command)
```
Comment (RodEsp) [inline src/help/index.ts:207]: Do we really need a whole new method for this if this is the only place it's used?

**Review by RodEsp — APPROVED** [review]

File: src/help/index.ts (hunk 7 above)
Comment (peternhale) [reply src/help/index.ts:207]: @RodEsp This function will be overridden in the custom help class to allow the determination of short help.

**Pull request merged**